File: src/main/java/cs/android/view/CSPagerAdapter.java
package cs.android.view;

import android.support.v4.view.PagerAdapter;
import android.view.View;
import android.view.ViewGroup;

import cs.android.viewbase.CSViewController;
import cs.java.collections.CSList;

import static cs.java.lang.CSLang.list;

public class CSPagerAdapter<T extends CSViewController & CSPagerPage> extends PagerAdapter {

    private CSList<T> _controllers = list();

    public CSPagerAdapter(CSList<T> controllers) {
        _controllers = controllers;
    }

    public void destroyItem(ViewGroup container, int position, Object object) {
        container.removeView((View) object);
    }

    public int getCount() {
        return _controllers.count();
    }

    public CharSequence getPageTitle(int position) {
        return _controllers.at(position).csPagerControllerTitle();
    }

    public Object instantiateItem(ViewGroup container, int position) {
        View view = _controllers.at(position).asView();
        container.addView(view, 0);
        return view;
    }

    public boolean isViewFromObject(View view, Object object) {
        return view == object;
    }

}

File: library/src/main/java/renetik/android/view/pager/CSPagerAdapter.kt
package renetik.android.view.pager

import android.view.View
import android.view.ViewGroup
import androidx.viewpager.widget.PagerAdapter
import renetik.java.collections.CSList
import renetik.java.collections.list
import renetik.android.view.base.CSViewController

class CSPagerAdapter<PagerPageType>(val controllers: CSList<PagerPageType> = list<PagerPageType>())
    : PagerAdapter() where PagerPageType : CSViewController<*>, PagerPageType : CSPagerPage {

    override fun destroyItem(container: ViewGroup, position: Int, `object`: Any) =
            container.removeView(`object` as View)

    override fun getCount() = controllers.size

    override fun getPageTitle(position: Int) = controllers[position].pagerPageTitle

    override fun instantiateItem(container: ViewGroup, position: Int): Any {
        val view = controllers[position].view
        container.addView(view, 0)
        return view
    }

    override fun isViewFromObject(view: View, `object`: Any) = view === `object`

}
